neps/nep-0413.md

```diff
@@ -0,0 +1,84 @@
+
+###  The problem with a restricted access key
+
+Originally NEAR was designed with the assumption that access keys would only be used to sign transactions. Using access keys for signing arbitrary payloads carries the risk of the signer inadvertently signing surreptitious transactions which the requester could in turn submit to the network.
+
+When using the keys for authentication, another issue that needs to be handled is a replay attack: a malicious app might request the user to authenticate only to authenticate itself to some other app as the user.
+
+If the wallet allows apps to request the user to sign arbitrary data, both attacks would be possible. Thus, some measures are needed to prevent both attacks:
+
+1. To prevent the malicious signing of transactions, a good solution is to make sure that the data to be signed cannot be parsed as a transaction. Having a standard format for the verification payload(different from the transaction format) will solve this.
+
+2. To prevent one app from relaying a signature to a different app, the data to be signed should include the requesting app's identity (e.g. domain name). Again, a well-defined format will facilitate this.
+
+Using an access key restricted in such a way that it cannot send transactions looks like a good temporary solution. It doesn't solve the second problem, though.
+
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead of a query string will improve this (the fragment isn't included in the Referrer), and somehow permitting only a fixed callback URL will improve this even more.
+
+Also, note that NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+Hence, the final solution we are settling on is to user a restricted payload as shown below.
+
+## Specification
+
+### `verifyOwner`
+
+Signs the message and verifies the owner. Message is not sent to blockchain.
+
+```jsx
+interface VerifyOwnerParams {
+  message?: string; // The message requested sign. Defaults to `verify owner` string.
+  callbackUrl?: string; // Applicable to browser wallets (e.g. MyNearWallet). This is the callback url once the signing is approved. Defaults to `window.location.href`.
+  meta?: string; // Applicable to browser wallets (e.g. MyNearWallet) extra data that will be passed to the callback url once the signing is approved.
+}
+
+interface VerifiedOwner {
+  accountId: string;
+  message: string;
+  blockId: string;
+  publicKey: string;
```

Comment:
Will it be possible to verify that indeed this `publicKey` belongs to this `accountId`?

Reply:
You would need to use the RPC API to know for sure, I think. https://docs.near.org/api/rpc/access-keys#view-access-key-list

Reply:
Not sure if the signing is done by an Access Key. Maybe the first Full Access Key (the one created at account creation) is used for signing? 

May be bit off topic, but verifying that the provided public key belongs to the account is equal amount important as verifying that the signature corresponds to the public key. The source of truth here is the blockchain.

Reply:
It should be a valid full access key, as far as I know. The wallet doing the verifying would use whatever is in their keystore for the requested account ID. I can't think of a way that an end user could know for sure that this was signed by a certain account otherwise.

I think the point of this method is to avoid the blockchain for simple verification of account ownership.

Reply:
You are correct, it is proposed that a newly created access key is to be used for ownership verification.

https://github.com/near/NEPs/blob/6a1403320ad6888065401216f4d56cb7503acd5c/neps/nep-0413.md?plain=1#L22 

> I think the point of this method is to avoid the blockchain for simple verification of account ownership.

Simple verification that can be **trusted**. If we cannot prove that the signature came from the public key owned by the account, anyone can create public key and sign the message.

```js
const ownershipResponse = await walletConnection.account().verifyOwnership({...})

const validSig = await verifySignature(ownershipResponse.signature, ownershipResponse.publicKey)
if (!validSig) {
  throw new Error('invalid signature for the public key, possibly tempered')
}

const trustedAccessKeys = await walletConnection.account().getAccessKeys()
if (!trustedAccessKeys.some(key => isPublicKeyEqual(key, ownershipResponse.publucKey)) {
  throw new Error('the key in the signature cannot be trusted, not found in the account access keys')
} 
```

Reply:
@toteto Checking if a full-access key belongs to a user is simple, check this [example I am working on](https://github.com/gagdiez/near-login/blob/main/server/authenticate/wallet-authenticate.js#L39-#L60).

The point of using a full-access key is that nobody should have access to your account's full-access key, except you. Furthermore, all activity involving a full-access key requires the user to accept it, therefore, the user would be notified of the signature happening, and would be required to accept it.